catalog/spatial/registry/registry-federation-admin-service-impl/src/main/java/org/codice/ddf/registry/federationadmin/service/impl/RegistryPublicationServiceImpl.java

```diff
@@ -0,0 +1,211 @@
+            metacard.setAttribute(new AttributeImpl(RegistryObjectMetacardType.LAST_PUBLISHED,
+                    Date.from(ZonedDateTime.now()
+                            .toInstant())));
+
+            federationAdminService.updateRegistryEntry(metacard);
+        }
+
+    }
+
+    @Override
+    public void unpublish(String registryId, String destinationRegistryId)
+            throws FederationAdminException {
+        Metacard metacard = getMetacard(registryId);
+
+        Attribute locAttr = metacard.getAttribute(RegistryObjectMetacardType.PUBLISHED_LOCATIONS);
+        if (locAttr == null || !locAttr.getValues()
+                .contains(destinationRegistryId)) {
+            return;
+        }
+
+        String sourceId = getSourceIdFromRegistryId(destinationRegistryId).orElse(null);
+
+        if (StringUtils.isNotBlank(sourceId)) {
+            federationAdminService.deleteRegistryEntriesByRegistryIds(Collections.singletonList(
+                    registryId), Collections.singleton(sourceId));
+
+            locAttr.getValues()
+                    .remove(destinationRegistryId);
+            metacard.setAttribute(locAttr);
+            metacard.setAttribute(new AttributeImpl(RegistryObjectMetacardType.LAST_PUBLISHED,
+                    Date.from(ZonedDateTime.now()
+                            .toInstant())));
+
+            federationAdminService.updateRegistryEntry(metacard);
+        }
+    }
+
+    @Override
+    public void update(Metacard metacard) throws FederationAdminException {
+        Attribute publishedLocations =
+                metacard.getAttribute(RegistryObjectMetacardType.PUBLISHED_LOCATIONS);
+        if (publishedLocations == null) {
+            return;
+        }
+
+        Set<String> locations = publishedLocations.getValues()
+                .stream()
+                .map(Object::toString)
+                .collect(Collectors.toCollection(HashSet::new));
+
+        if (CollectionUtils.isNotEmpty(locations)) {
+            try {
+                federationAdminService.updateRegistryEntry(metacard, locations);
+                metacard.setAttribute(new AttributeImpl(RegistryObjectMetacardType.LAST_PUBLISHED,
+                        Date.from(ZonedDateTime.now()
+                                .toInstant())));
+                federationAdminService.updateRegistryEntry(metacard);
+            } catch (FederationAdminException e) {
+                LOGGER.warn("Unable to send update for {} to {}", metacard.getTitle(), locations);
+            }
+        }
+    }
+
+    public void bindRegistryStore(ServiceReference serviceReference) {
+        BundleContext bundleContext = getBundleContext();
+
+        if (serviceReference != null && bundleContext != null) {
+            RegistryStore registryStore =
+                    (RegistryStore) bundleContext.getService(serviceReference);
+
+            registryStoresByRegistryId.put(registryStore.getRegistryId(), registryStore);
```

Comment:
Unfortunately the stores registry id might be an empty string when the service first comes up since it has to connect to the remote registry and get the registry id from it. So you could be overwriting your stores since multiple stores will initially have a registry id of ''. Recommend changing this to a list of RegistryStores that you loop through to find the right store when you need it.